File: android/app/src/main/kotlin/com/musicmuni/calibra/sample/SingalongPracticeSection.kt
package com.musicmuni.calibra.sample

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.musicmuni.vozos.calibra.CalibraPitch
import com.musicmuni.vozos.calibra.CalibraSegmentEvaluator
import com.musicmuni.vozos.sonix.SonixDecoder
import com.musicmuni.vozos.sonix.SonixPlayer
import com.musicmuni.vozos.sonix.SonixRecorder
import com.musicmuni.vozos.sonix.SonixResampler
import com.musicmuni.vozos.sonix.util.Parser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * Singalong Practice Section - for lessons like Alankaar where user sings along with teacher.
 *
 * Demonstrates:
 * - `CalibraSegmentEvaluator` for sing-along evaluation
 * - `CalibraPitch` for real-time pitch detection
 * - `SonixRecorder` for audio capture
 *
 * Flow:
 * 1. Load lesson assets (.trans, .mp3)
 * 2. User selects a segment
 * 3. Play reference audio while user sings along
 * 4. Capture user's pitch in real-time
 * 5. Evaluate and show score after segment ends
 */
@Composable
fun SingalongPracticeSection() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    // Lesson state
    var lessonLoaded by remember { mutableStateOf(false) }
    var segments by remember { mutableStateOf<List<SegmentInfo>>(emptyList()) }
    var currentSegmentIndex by remember { mutableStateOf(0) }
    val lessonName = "Alankaar 01"

    // Playback state
    var player by remember { mutableStateOf<SonixPlayer?>(null) }
    var isPlaying by remember { mutableStateOf(false) }
    var currentTimeMs by remember { mutableLongStateOf(0L) }

    // Recording state
    var recorder by remember { mutableStateOf<SonixRecorder?>(null) }
    var isRecording by remember { mutableStateOf(false) }
    var recordingLevel by remember { mutableFloatStateOf(0f) }

    // Calibra public API
    var evaluator by remember { mutableStateOf<CalibraSegmentEvaluator?>(null) }
    var pitchDetector by remember { mutableStateOf<CalibraPitch?>(null) }

    var currentPitch by remember { mutableFloatStateOf(0f) }
    var currentNote by remember { mutableStateOf("-") }

    // Evaluation state
    var segmentScore by remember { mutableFloatStateOf(0f) }
    var practiceState by remember { mutableStateOf(PracticeState.IDLE) }
    var status by remember { mutableStateOf("Ready") }
    var segmentEndTimeMs by remember { mutableLongStateOf(Long.MAX_VALUE) }

    // Copy asset to cache and return path
    suspend fun copyAssetToCache(assetName: String): String {
        return withContext(Dispatchers.IO) {
            val cacheFile = File(context.cacheDir, assetName)
            if (!cacheFile.exists()) {
                context.assets.open(assetName).use { input ->
                    FileOutputStream(cacheFile).use { output ->
                        input.copyTo(output)
                    }
                }
            }
            cacheFile.absolutePath
        }
    }

    // Load lesson
    fun loadLesson() {
        scope.launch {
            status = "Loading lesson..."
            try {
                // Create pitch detector using Calibra public API (with processing for smoothing + octave correction)
                pitchDetector?.release()
                pitchDetector = CalibraPitch.create(enableProcessing = true)

                // Parse trans file
                val transContent = withContext(Dispatchers.IO) {
                    context.assets.open("$lessonName.trans").bufferedReader().readText()
                }
                val transData = Parser.parseTransString(transContent)
                if (transData == null) {
                    status = "Failed to parse trans file"
                    return@launch
                }

                // Convert to SegmentInfo
                segments = transData.segments.mapIndexed { index, seg ->
                    SegmentInfo(
                        index = index,
                        lyrics = seg.lyrics,
                        startTime = seg.startTime,
                        endTime = seg.endTime
                    )
                }

                // Copy audio to cache and create player
                val audioPath = copyAssetToCache("$lessonName.mp3")
                player?.release()
                player = SonixPlayer.create(audioPath)

                // Create recorder for capturing user audio
                val recordPath = "${context.cacheDir}/user_recording.m4a"
                recorder?.release()
                recorder = SonixRecorder.create(recordPath, "m4a", "voice")

                // Decode reference audio using Sonix and create evaluator
                val audioData = withContext(Dispatchers.IO) {
                    SonixDecoder.decode(audioPath)
                }
                if (audioData != null) {
                    // Resample reference audio to 16kHz
                    val refSamples16k = SonixResampler.resample(
                        samples = audioData.floatSamples,
                        fromRate = audioData.sampleRate,
                        toRate = 16000
                    )
                    // Create evaluator from audio reference at 16kHz
                    evaluator?.release()
                    evaluator = CalibraSegmentEvaluator.createFromAudio(
                        refSamplesMono = refSamples16k,
                        sampleRate = 16000
                    )
                } else {
                    status = "Failed to decode reference audio"
                    return@launch
                }

                lessonLoaded = true
                currentSegmentIndex = 0
                status = "Lesson loaded: ${segments.size} segments"
            } catch (e: Exception) {
                status = "Error: ${e.message}"
            }
        }
    }

    // Start practicing a segment
    fun startPractice() {
        if (!lessonLoaded || segments.isEmpty()) return

        scope.launch {
            val segment = segments[currentSegmentIndex]
            practiceState = PracticeState.PRACTICING
            segmentScore = 0f
            status = "Practicing segment ${currentSegmentIndex + 1}..."

            // Precompute reference features for this segment using Calibra public API
            evaluator?.precomputeRef(
                segmentIndex = currentSegmentIndex,
                refStart = segment.startTime.toFloat(),
                refEnd = segment.endTime.toFloat()
            )

            // Seek player to segment start
            player?.seek((segment.startTime * 1000).toLong())
            segmentEndTimeMs = (segment.endTime * 1000).toLong()

            // Start capturing using Calibra public API
            evaluator?.startCapture(segmentIndex = currentSegmentIndex)

            // Start recording and playback
            recorder?.start()
            player?.play()
            isPlaying = true
            isRecording = true

            // Collect audio buffers for pitch detection
            launch {
                val hwRate = 48000 // Android hardware sample rate

                recorder?.audioBuffers?.collect { buffer ->
                    if (!isRecording) return@collect

                    // Prepare audio for Calibra (resample to 16kHz)
                    val floatBuffer = FloatArray(buffer.sampleCount)
                    buffer.fillFloatSamples(floatBuffer)
                    val samples = SonixResampler.resample(
                        samples = floatBuffer,
                        fromRate = hwRate,
                        toRate = 16000
                    )

                    // Use pitch detector directly for real-time pitch display
                    val result = pitchDetector?.detectWithConfidence(samples = samples)
                    val detectedPitch = result?.pitchHz ?: -1.0f
                    val calculatedRms = pitchDetector?.getAmplitude(samples = samples) ?: 0.0f

                    // Also feed to evaluator for scoring
                    evaluator?.processFrame(samples = samples)

                    currentPitch = detectedPitch
                    recordingLevel = calculatedRms
                    currentNote = MusicUtils.getMidiNoteName(detectedPitch)
                }
            }

            // Collect playback time
            launch {
                player?.currentTime?.collect { timeMs ->
                    currentTimeMs = timeMs
                }
            }
        }
    }

    // Auto-stop when segment ends
    LaunchedEffect(currentTimeMs, practiceState) {
        if (practiceState == PracticeState.PRACTICING && currentTimeMs >= segmentEndTimeMs && isPlaying) {
            val segment = segments[currentSegmentIndex]

            // Stop playback and recording
            player?.pause()
            recorder?.stop()
            isPlaying = false
            isRecording = false

            // Evaluate using Calibra public API
            evaluator?.stopCapture(segmentIndex = currentSegmentIndex)
            segmentScore = evaluator?.evaluate(
                segmentIndex = currentSegmentIndex,
                stdStart = segment.startTime.toFloat(),
                stdEnd = segment.endTime.toFloat()
            ) ?: 0f

            practiceState = PracticeState.EVALUATED
            status = "Segment ${currentSegmentIndex + 1} score: ${(segmentScore * 100).toInt()}%"
            segmentEndTimeMs = Long.MAX_VALUE
        }
    }

    // Stop practice and evaluate (manual stop)
    fun stopPractice() {
        if (practiceState != PracticeState.PRACTICING) return

        scope.launch {
            val segment = segments[currentSegmentIndex]

            // Stop playback and recording
            player?.pause()
            recorder?.stop()
            isPlaying = false
            isRecording = false

            // Evaluate using Calibra public API
            evaluator?.stopCapture(segmentIndex = currentSegmentIndex)
            segmentScore = evaluator?.evaluate(
                segmentIndex = currentSegmentIndex,
                stdStart = segment.startTime.toFloat(),
                stdEnd = segment.endTime.toFloat()
            ) ?: 0f

            practiceState = PracticeState.EVALUATED
            status = "Segment ${currentSegmentIndex + 1} score: ${(segmentScore * 100).toInt()}%"
        }
    }

    // Cleanup on dispose
    DisposableEffect(Unit) {
        onDispose {
            player?.release()
            recorder?.release()
            pitchDetector?.release()
            evaluator?.release()
        }
    }

    // UI
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "Singalong Practice (Alankaar)",
            style = MaterialTheme.typography.titleMedium
        )

        Text(
            text = status,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        // Load button
        if (!lessonLoaded) {
            Button(onClick = { loadLesson() }) {
                Text("Load Lesson")
            }
        } else {
            // Segment selector
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Segment ${currentSegmentIndex + 1} of ${segments.size}",
                    style = MaterialTheme.typography.bodyMedium
                )

                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Button(
                        onClick = { if (currentSegmentIndex > 0) currentSegmentIndex-- },
                        enabled = currentSegmentIndex > 0 && practiceState != PracticeState.PRACTICING
                    ) {
                        Text("<")
                    }
                    Button(
                        onClick = { if (currentSegmentIndex < segments.size - 1) currentSegmentIndex++ },
                        enabled = currentSegmentIndex < segments.size - 1 && practiceState != PracticeState.PRACTICING
                    ) {
                        Text(">")
                    }
                }
            }

            // Current segment lyrics
            if (segments.isNotEmpty()) {
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.primaryContainer
                    )
                ) {
                    Text(
                        text = segments[currentSegmentIndex].lyrics,
                        style = MaterialTheme.typography.bodyLarge,
                        modifier = Modifier.padding(12.dp)
                    )
                }
            }

            // Practice controls
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                when (practiceState) {
                    PracticeState.IDLE, PracticeState.EVALUATED -> {
                        Button(
                            onClick = { startPractice() },
                            modifier = Modifier.weight(1f)
                        ) {
                            Text("Start Practice")
                        }
                    }
                    PracticeState.PRACTICING -> {
                        Button(
                            onClick = { stopPractice() },
                            modifier = Modifier.weight(1f),
                            colors = ButtonDefaults.buttonColors(
                                containerColor = MaterialTheme.colorScheme.error
                            )
                        ) {
                            Text("Stop")
                        }
                    }
                }
            }

            // Real-time feedback during practice
            if (practiceState == PracticeState.PRACTICING) {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    // Pitch display
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween
                    ) {
                        Text("Pitch: ${"%.1f".format(currentPitch)} Hz")
                        Text(
                            text = currentNote,
                            style = MaterialTheme.typography.titleLarge
                        )
                    }

                    // Level meter
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Text("Level:", style = MaterialTheme.typography.bodySmall)
                        LinearProgressIndicator(
                            progress = { recordingLevel.coerceIn(0f, 1f) },
                            modifier = Modifier
                                .weight(1f)
                                .height(8.dp)
                        )
                    }
                }
            }

            // Score display after evaluation
            if (practiceState == PracticeState.EVALUATED) {
                ScoreCard(score = segmentScore)
            }
        }
    }
}

data class SegmentInfo(
    val index: Int,
    val lyrics: String,
    val startTime: Double,
    val endTime: Double
)

enum class PracticeState {
    IDLE,
    PRACTICING,
    EVALUATED
}

@Composable
fun ScoreCard(score: Float) {
    val backgroundColor = when {
        score >= 0.8f -> Color(0xFF4CAF50)
        score >= 0.6f -> Color(0xFFFF9800)
        else -> Color(0xFFF44336)
    }
    val feedback = when {
        score >= 0.8f -> "Excellent!"
        score >= 0.6f -> "Good job!"
        else -> "Keep practicing!"
    }

    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = backgroundColor)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                text = "${(score * 100).toInt()}%",
                style = MaterialTheme.typography.headlineLarge,
                color = Color.White
            )
            Text(
                text = feedback,
                color = Color.White
            )
        }
    }
}

object MusicUtils {
    fun getMidiNoteName(pitchHz: Float): String {
        if (pitchHz <= 0) return "-"
        val midiNote = (12 * kotlin.math.log2(pitchHz / 440.0) + 69).toInt()
        val noteNames = arrayOf("C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B")
        val noteName = noteNames[midiNote % 12]
        val octave = (midiNote / 12) - 1
        return "$noteName$octave"
    }
}

File: ios/CalibraSample/CalibraSample/Sections/SingalongPracticeSection.swift
import SwiftUI
import calibra
import sonix

enum PracticeState {
    case idle
    case practicing
    case evaluated
}

/// Singalong Practice Section - for lessons like Alankaar where user sings along with teacher.
///
/// Demonstrates:
/// - `CalibraRealtimeSession` for real-time singing evaluation
/// - `SonixRecorder` for audio capture
/// - `SonixPlayer` for reference audio playback
///
/// Flow:
/// 1. Load lesson assets (.trans, .mp3)
/// 2. User selects a segment
/// 3. Play reference audio while user sings along
/// 4. Capture user's pitch in real-time
/// 5. Evaluate and show score after segment ends
struct SingalongPracticeSection: View {
    // Lesson state
    @State private var lessonLoaded = false
    @State private var currentSegmentIndex = 0
    @State private var lessonName = "Alankaar 01"

    // Playback state (using SonixPlayer)
    @State private var player: SonixPlayer?
    @State private var isPlaying = false
    @State private var currentTimeMs: Int64 = 0
    @State private var playerObserverTask: Task<Void, Never>?

    // Recording state
    @State private var isRecording = false
    @State private var recordingLevel: Float = 0.0

    // Sonix recorder
    @State private var recorder: SonixRecorder?

    // Calibra new API - CalibraRealtimeSession
    @State private var session: CalibraRealtimeSession?
    @State private var pitchDetector: CalibraPitch?

    @State private var currentPitch: Float = 0.0
    @State private var currentNote = "-"

    // Evaluation state
    @State private var segmentScore: Float = 0.0
    @State private var practiceState: PracticeState = .idle
    @State private var status = "Ready"
    @State private var feedbackMessage = ""

    // Timer for checking segment end
    @State private var timer: Timer?
    @State private var segmentEndTime: Double = 0

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Singalong Practice (Alankaar)")
                .font(.headline)

            Text(status)
                .font(.caption)
                .foregroundColor(.secondary)

            if !lessonLoaded {
                Button("Load Lesson") {
                    loadLesson()
                }
                .buttonStyle(.borderedProminent)
            } else if let session = session {
                // Segment selector
                let segments = session.segments
                HStack {
                    Text("Segment \(currentSegmentIndex + 1) of \(segments.count)")
                        .font(.subheadline)

                    Spacer()

                    HStack(spacing: 8) {
                        Button("<") {
                            if currentSegmentIndex > 0 {
                                currentSegmentIndex -= 1
                            }
                        }
                        .disabled(currentSegmentIndex == 0 || practiceState == .practicing)

                        Button(">") {
                            if currentSegmentIndex < segments.count - 1 {
                                currentSegmentIndex += 1
                            }
                        }
                        .disabled(currentSegmentIndex >= segments.count - 1 || practiceState == .practicing)
                    }
                    .buttonStyle(.bordered)
                }

                // Current segment lyrics
                if !segments.isEmpty {
                    Text(segments[currentSegmentIndex].lyrics)
                        .font(.body)
                        .padding(12)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color.blue.opacity(0.1))
                        .cornerRadius(8)
                }

                // Practice controls
                HStack {
                    switch practiceState {
                    case .idle, .evaluated:
                        Button("Start Practice") {
                            startPractice()
                        }
                        .buttonStyle(.borderedProminent)
                        .frame(maxWidth: .infinity)

                    case .practicing:
                        Button("Stop") {
                            stopPractice()
                        }
                        .buttonStyle(.borderedProminent)
                        .tint(.red)
                        .frame(maxWidth: .infinity)
                    }
                }

                // Real-time feedback during practice
                if practiceState == .practicing {
                    VStack(spacing: 8) {
                        HStack {
                            Text("Pitch: \(String(format: "%.1f", currentPitch)) Hz")
                            Spacer()
                            Text(currentNote)
                                .font(.title2)
                                .fontWeight(.bold)
                        }

                        HStack {
                            Text("Level:")
                                .font(.caption)
                            ProgressView(value: Double(min(max(recordingLevel, 0), 1)))
                        }
                    }
                }

                // Score display after evaluation
                if practiceState == .evaluated {
                    scoreCard
                }
            }
        }
        .onDisappear {
            cleanup()
        }
    }

    private var scoreCard: some View {
        let backgroundColor: Color = {
            if segmentScore >= 0.8 {
                return .green
            } else if segmentScore >= 0.6 {
                return .orange
            } else {
                return .red
            }
        }()

        return VStack(spacing: 8) {
            Text("\(Int(segmentScore * 100))%")
                .font(.largeTitle)
                .fontWeight(.bold)
                .foregroundColor(.white)
            Text(feedbackMessage)
                .foregroundColor(.white)
        }
        .frame(maxWidth: .infinity)
        .padding(16)
        .background(backgroundColor)
        .cornerRadius(12)
    }

    private func setupAudioIfNeeded() {
        guard pitchDetector == nil else { return }

        // Create pitch detector using Calibra public API (with processing for smoothing + octave correction)
        pitchDetector = CalibraPitch.create(enableProcessing: true)

        // Create recorder using Sonix with echo cancellation enabled
        // This removes the lesson audio from the mic input so pitch detection works correctly
        let tempPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("singalong_temp.m4a").path
        recorder = SonixRecorder.create(outputPath: tempPath, format: "m4a", quality: "voice", echoCancellation: true)
    }

    private func cleanup() {
        timer?.invalidate()
        playerObserverTask?.cancel()
        player?.stop()
        player?.release()
        player = nil
        recorder?.stop()
        recorder?.release()
        recorder = nil

        // Clean up Calibra objects
        pitchDetector?.release()
        pitchDetector = nil
        session?.close()
        session = nil
    }

    private func loadLesson() {
        setupAudioIfNeeded()
        status = "Loading lesson..."

        // Load and parse trans file from bundle
        guard let transURL = Bundle.main.url(forResource: lessonName, withExtension: "trans"),
              let transContent = try? String(contentsOf: transURL),
              let transData = Parser.parseTransString(content: transContent) else {
            status = "Failed to load lesson - trans file not found"
            return
        }

        // Convert TransData segments to Calibra Segment model
        let segments: [Segment] = transData.segments.enumerated().map { (index, seg) in
            Segment(
                index: Int32(index),
                startSeconds: Float(seg.startTime),
                endSeconds: Float(seg.endTime),
                lyrics: seg.lyrics
            )
        }

        // Load audio file from bundle (m4a format)
        guard let audioURL = Bundle.main.url(forResource: lessonName, withExtension: "m4a") else {
            status = "Failed to load lesson - audio file not found"
            return
        }

        // Load player using Sonix (async)
        Task {
            do {
                player = try await SonixPlayer.create(source: audioURL.path)

                // Observe current time for segment end detection
                playerObserverTask = player?.observeCurrentTime { timeMs in
                    self.currentTimeMs = timeMs
                }
            } catch {
                await MainActor.run {
                    status = "Error loading audio: \(error)"
                }
                return
            }

            // Decode reference audio using Sonix and create session with new API
            if let audioData = SonixDecoder.decode(path: audioURL.path) {
                print("[DEBUG] Decoded audio: \(audioData.floatSamples.size) samples at \(audioData.sampleRate) Hz")

                // Create SingingReference using new API
                // Resampling is handled internally by CalibraRealtimeSession
                let reference = SingingReference.fromAudio(
                    samples: audioData.floatSamples,
                    sampleRate: audioData.sampleRate,
                    segments: segments,
                    keyHz: 196.0  // G3, common for Indian classical
                )

                // Create session with config (manual segment control)
                let config = SessionConfig.manual
                session = CalibraRealtimeSession.create(reference: reference, config: config)

                // Prepare session (precomputes reference features)
                session?.prepare()
            } else {
                await MainActor.run {
                    status = "Failed to decode reference audio"
                }
                return
            }

            await MainActor.run {
                lessonLoaded = true
                currentSegmentIndex = 0
                status = "Lesson loaded: \(segments.count) segments"
            }
        }
    }

    private func startPractice() {
        guard lessonLoaded, let session = session, let player = player else { return }

        let segments = session.segments
        guard currentSegmentIndex < segments.count else { return }

        let segment = segments[currentSegmentIndex]
        practiceState = .practicing
        segmentScore = 0
        feedbackMessage = ""
        status = "Practicing segment \(currentSegmentIndex + 1)..."

        // Seek player to segment start and play (convert seconds to milliseconds)
        let startMs = Int64(segment.startSeconds * 1000)
        segmentEndTime = Double(segment.endSeconds)
        player.seek(positionMs: startMs)
        player.play()
        isPlaying = true

        // Begin segment using new CalibraRealtimeSession API
        session.beginSegment(index: Int32(currentSegmentIndex))

        // Start recording
        startRecording()

        // Start timer to check for segment end (using observed currentTimeMs)
        timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { _ in
            let currentTimeSec = Double(currentTimeMs) / 1000.0
            if currentTimeSec >= segmentEndTime {
                stopPractice()
            }
        }
    }

    private func stopPractice() {
        timer?.invalidate()
        timer = nil

        player?.pause()
        isPlaying = false

        stopRecording()

        // End segment early and get result using new API
        if let result = session?.endSegmentEarly() {
            segmentScore = result.score
            feedbackMessage = result.feedbackMessage
        }

        practiceState = .evaluated
        status = "Segment \(currentSegmentIndex + 1) score: \(Int(segmentScore * 100))%"
    }

    private func startRecording() {
        guard let recorder = recorder else { return }

        isRecording = true

        // Collect audio buffers from Sonix
        var frameCount = 0
        Task {
            let hwRate = Int(Sonix.hardwareSampleRate)

            for await buffer in recorder.audioBuffers {
                frameCount += 1

                // Resample to 16kHz for Calibra (expects 16kHz input)
                let samples16k = SonixResampler.resample(
                    samples: buffer.floatSamples,
                    fromRate: hwRate,
                    toRate: 16000
                )

                // Feed resampled audio to session
                session?.addAudio(samples: samples16k)

                // Use pitch detector for real-time pitch display
                let result = pitchDetector?.detectWithConfidence(samples: samples16k)
                let detectedPitch = result?.pitchHz ?? -1.0
                let calculatedRms = pitchDetector?.getAmplitude(samples: samples16k) ?? 0.0

                // Log first few frames and every 50th frame
                if frameCount <= 3 || frameCount % 50 == 0 {
                    print("[DEBUG] Frame \(frameCount): \(samples16k.size) samples, pitch: \(detectedPitch) Hz")
                }

                await MainActor.run {
                    currentPitch = detectedPitch
                    recordingLevel = calculatedRms
                    currentNote = MusicUtils.getMidiNoteName(detectedPitch)
                }
            }
        }

        recorder.start()
    }

    private func stopRecording() {
        recorder?.stop()
        isRecording = false
    }
}
